hail/src/main/scala/is/hail/expr/ir/functions/Functions.scala

```diff
@@ -395,147 +370,141 @@ abstract class RegistryFunctions {
     registerWrappedScalaFunction(mname, Array(a1, a2, a3), rType, unwrappedApply(pt))(cls, method)
 
   def registerJavaStaticFunction(mname: String, argTypes: Array[Type], rType: Type, pt: (Type, Seq[PType]) => PType)(cls: Class[_], method: String) {
-    registerCode(mname, argTypes, rType, pt) { case (r, rt, args) =>
+    registerCode(mname, argTypes, rType, pt) { case (r, rt, _, args) =>
       val cts = argTypes.map(TypeToIRIntermediateClassTag(_).runtimeClass)
       Code.invokeStatic(cls, method, cts, args.map(_._2))(TypeToIRIntermediateClassTag(rType))
     }
   }
 
-  def registerIR(mname: String, typeArgs: Array[Type], argTypes: Array[Type], retType: Type, inline: Boolean)(f: (Seq[Type], Seq[IR]) => IR): Unit =
-    IRFunctionRegistry.addIR(mname, typeArgs, argTypes, retType, inline, f)
+  def registerIR(mname: String, argTypes: Array[Type], retType: Type, inline: Boolean = false, typeParams: Array[Type] = Array.empty)(f: (Seq[Type], Seq[IR]) => IR): Unit =
+    IRFunctionRegistry.addIR(mname, typeParams, argTypes, retType, inline, f)
 
-  def registerIR(mname: String, argTypes: Array[Type], retType: Type, inline: Boolean = false)(f: (Seq[Type], Seq[IR]) => IR): Unit = {
-    registerIR(mname, Array.empty[Type], argTypes, retType, inline)(f)
-  }
 
-  def registerPCode(mname: String, mt1: Type, rt: Type, pt: (Type, PType) => PType)(impl: (EmitRegion, PType, PCode) => PCode): Unit =
+  def registerPCode1(mname: String, mt1: Type, rt: Type, pt: (Type, PType) => PType)(impl: (EmitRegion, PType, PCode) => PCode): Unit =
     registerPCode(mname, Array(mt1), rt, unwrappedApply(pt)) {
       case (r, rt, Array(a1)) => impl(r, rt, a1)
     }
 
-  def registerPCode(mname: String, mt1: Type, mt2: Type, rt: Type, pt: (Type, PType, PType) => PType)
+  def registerPCode2(mname: String, mt1: Type, mt2: Type, rt: Type, pt: (Type, PType, PType) => PType)
     (impl: (EmitRegion, PType, PCode, PCode) => PCode): Unit =
     registerPCode(mname, Array(mt1, mt2), rt, unwrappedApply(pt)) {
       case (r, rt, Array(a1, a2)) => impl(r, rt, a1, a2)
     }
 
-  def registerCode[A1](mname: String, mt1: Type, rt: Type, pt: (Type, PType) => PType)(impl: (EmitRegion, PType, (PType, Code[A1])) => Code[_]): Unit =
+  def registerCode1[A1](mname: String, mt1: Type, rt: Type, pt: (Type, PType) => PType)(impl: (EmitRegion, PType, (PType, Code[A1])) => Code[_]): Unit =
     registerCode(mname, Array(mt1), rt, unwrappedApply(pt)) {
-      case (r, rt, Array(a1: (PType, Code[A1]) @unchecked)) => impl(r, rt, a1)
+      case (r, rt, _, Array(a1: (PType, Code[A1]) @unchecked)) => impl(r, rt, a1)
     }
 
-  def registerCode[A1, A2](mname: String, mt1: Type, mt2: Type, rt: Type, pt: (Type, PType, PType) => PType)
+  def registerCode1t[A1](mname: String, typeArg: Type, mt1: Type, rt: Type, pt: (Type, PType) => PType)(impl: (EmitRegion, PType, Type, (PType, Code[A1])) => Code[_]): Unit =
+    registerCode(mname, Array(mt1), rt, unwrappedApply(pt)) {
```

Comment:
registerCode should take Array(typeArg) in the last position.

Reply:
can also be named typeArg1 for consistency

Reply:
and we should probably name all of these either typeArg as before, or typeParam as you did above.

Reply:
good catch, thanks

Reply:
yeah, renamed to typeParams everywhere. I think it's a bit clearer not to have argTypes and typeArgs 😂